modules/ROOT/pages/patterns/shortest-paths.adoc

```diff
@@ -0,0 +1,540 @@
+| ["Worcester Shrub Hill", "Ashchurch", "Cheltenham Spa", "Bromsgrove"]
+
+1+d|Rows: 5
+|===
+
+If there had been only four possible paths between the two Stations, then only those four would have been returned.
+
+[[all-shortest]]
+== `ALL SHORTEST`
+
+To return all paths that are tied for shortest length, use the keywords `ALL SHORTEST`:
+
+.Query
+[source,cypher]
+----
+MATCH p = ALL SHORTEST (wos)-[:LINK]-+(cnm)
+WHERE hby.name = "Worcester Shrub Hill" AND cnm.name = "Bromsgrove"
+RETURN [n in nodes(p) | n.name] AS stops
+----
+
+.Result
+[role="queryresult",options="header,footer",cols="m"]
+|===
+| stops
+
+| ["Worcester Shrub Hill", "Droitwich Spa", "Bromsgrove"]
+| ["Worcester Shrub Hill", "Worcestershire Parkway Ll", "Bromsgrove"]
+
+1+d|Rows: 2
+
+|===
+
+
+[[shortest-groups]]
+== `SHORTEST k GROUPS`
+
+To return all paths that are tied for first, second, and so on up to the kth shortest length, use `SHORTEST k GROUPS`.
+For example, the following returns the first and second shortest length paths between  `Worcester Shrub Hill` and `Bromsgrove`:
+
+.Query
+[source,cypher]
+----
+MATCH p = SHORTEST 2 GROUPS (wos:Station)-[:LINK]-+(bmv:Station)
+WHERE wos.name = "Worcester Shrub Hill" AND bmv.name = "Bromsgrove"
+RETURN [n in nodes(p) | n.name] AS stops, length(p) AS pathLength
+----
+
+.Result
+[role="queryresult",options="header,footer",cols="2m,m"]
+|===
+| stops | pathLength
+| ["Worcester Shrub Hill", "Droitwich Spa", "Bromsgrove"] | 2
+| ["Worcester Shrub Hill", "Worcestershire Parkway Ll", "Bromsgrove"] | 2
+| ["Worcester Shrub Hill", "Worcester Foregate Street", "Droitwich Spa", "Bromsgrove"] | 3
+| ["Worcester Shrub Hill", "Ashchurch", "Worcestershire Parkway Ll", "Bromsgrove"] | 3
+| ["Worcester Shrub Hill", "Ashchurch", "Cheltenham Spa", "Bromsgrove"] | 3
+2+d|Rows: 5
+|===
+
+The first group includes the two shortest paths with `pathLength = 2` (as seen in the first two rows of the results), and the second group includes the three second shortest paths with `pathLength = 3` (as seen in the last three rows of the results).
+
+If more groups are specified than exist in the graph, only those paths that exist are returned.
+For example, if the paths equal to one of the eight shortest paths are specified for `Worcester Shrub Hill` to `Bromsgrove`, only seven groups are returned:
+
+.Query
+[source,cypher]
+----
+MATCH p = SHORTEST 8 GROUPS (wos:Station)-[:LINK]-+(bmv:Station)
+WHERE wos.name = "Worcester Shrub Hill" AND bmv.name = "Bromsgrove"
+RETURN length(p) AS pathLength, count(*) AS numPaths
+----
+
+.Result
+[role="queryresult",options="header,footer",cols="2*m"]
+|===
+| pathLength | numPaths
+
+| 2 | 2
+| 3 | 3
+| 4 | 1
+| 5 | 4
+| 6 | 8
+| 7 | 10
+| 8 | 6
+
+2+d|Rows: 7
+|===
+
+[[any]]
+== `ANY`
+
+The `ANY` keyword can be used to test the reachability of nodes from a given node(s).
+It returns the same as `SHORTEST 1`, but by using the `ANY` keyword the intent of the query is clearer.
```

Comment:
Are we sure that we want to document that `ANY` returns the same as `SHORTEST 1`? In general, this is not a hard constraint right? I am just thinking what happens if we find in the future an algorithm which can efficiently find and return a path if it exists, but maybe not necessarily the same as `SHORTEST 1` finds. Also, `SHORTEST 1` returns one of the possibly many shortest paths non-deterministically.
I would choose for a description like this: _At the moment, `ANY` is rewritten to `SHORTEST 1`, but by using the `ANY` keyword the intent of the query is clearer._

Reply:
It originally read:

> Currently it is implemented to return one of the shortest paths between the node(s), but the keyword can be used to more clearly express the intent of the query

but the style guidance from @JPryce-Aklundh was to avoid statements like this and rather update it when and if such a change is made.

Reply:
Yes, @WilcoNeo - we prefer to document current functionality and update documentation whenever that changes.